diff --git a/src/workspace.c b/src/workspace.c
index 8bd4b5e..29bfc15 100644
--- a/src/workspace.c
+++ b/src/workspace.c
@@ -653,6 +653,19 @@ Con *workspace_next_on_output(void) {
         }
     }

+    /* Find first workspace on next output */
+    if (!next) {
+        Output *current_output = get_output_by_name(output->name);
+        Output *next_output = get_output_from_string(current_output, "right");
+
+        NODES_FOREACH(output_get_content(next_output->con)) {
+            if (child->type != CT_WORKSPACE)
+                continue;
+            if (!next || (child->num != -1 && child->num < next->num))
+                next = child;
+        }
+    }
+
     /* Find first workspace. */
     if (!next) {
         NODES_FOREACH(output_get_content(output)) {
@@ -694,6 +707,24 @@ Con *workspace_prev_on_output(void) {
         }
     }

+    /* Find last workspace on prev output if no previous numbered */
+
+    if (current->num > -1 && !prev) {
+        Output *current_output = get_output_by_name(output->name);
+        Output *prev_output = get_output_from_string(current_output, "left");
+
+        NODES_FOREACH_REVERSE(output_get_content(prev_output->con)) {
+            if (child->type != CT_WORKSPACE)
+                continue;
+            if (!prev || child->num > prev->num)
+                prev = child;
+            if (child->num == -1) {
+                prev = child;
+                goto workspace_prev_on_output_end;
+            }
+        }
+    }
+
     /* Find previous named workspace. */
     if (!prev) {
         bool found_current = false;
